vdr/vdr.go

```diff
@@ -0,0 +1,182 @@
+}
+
+var instance *Registry
+var oneRegistry sync.Once
+
+// ReloadRegistryIdleTimeout defines the cooling down period after receiving a file watcher notification, before
+// the registry is reloaded (from disk).
+var ReloadRegistryIdleTimeout time.Duration
+
+func init() {
+	ReloadRegistryIdleTimeout = 3 * time.Second
+}
+
+// RegistryInstance returns the singleton Registry
+func RegistryInstance() *Registry {
+	if instance != nil {
+		return instance
+	}
+	oneRegistry.Do(func() {
+		instance = NewRegistryInstance(DefaultRegistryConfig(), crypto.Instance(), networkClient.NewNetworkClient())
+	})
+
+	return instance
+}
+
+func NewRegistryInstance(config Config, cryptoClient crypto.KeyStore, networkClient pkg.NetworkClient) *Registry {
+	return &Registry{
+		Config:        config,
+		crypto:        cryptoClient,
+		network:       networkClient,
+		_logger:       logging.Log(),
+		store:         store.NewMemoryStore(),
+		didDocCreator: NutsDocCreator{keyCreator: cryptoClient},
+	}
+}
+
+// Configure initializes the db, but only when in server mode
+func (r *Registry) Configure() error {
+	var err error
+
+	r.configOnce.Do(func() {
+		cfg := core.NutsConfig()
+		r.Config.Mode = cfg.GetEngineMode(r.Config.Mode)
+		if r.Config.Mode == core.ServerEngineMode {
+			//r.Db = db.New()
+			if r.networkAmbassador == nil {
+				r.networkAmbassador = network.NewAmbassador(r.network, r.crypto)
+			}
+		}
+	})
+	return err
+}
+
+// Start initiates the routines for auto-updating the data
+func (r *Registry) Start() error {
+	if r.Config.Mode == core.ServerEngineMode {
+		r.networkAmbassador.Start()
+	}
+	return nil
+}
+
+// Shutdown cleans up any leftover go routines
+func (r *Registry) Shutdown() error {
+	if r.Config.Mode == core.ServerEngineMode {
+		logging.Log().Debug("Sending close signal to all routines")
+		for _, ch := range r.closers {
+			ch <- struct{}{}
+		}
+		logging.Log().Info("All routines closed")
+	}
+	return nil
+}
+
+func (r *Registry) Diagnostics() []core.DiagnosticResult {
+	return []core.DiagnosticResult{}
+}
+
+func (r *Registry) getEventsDir() string {
+	return r.Config.Datadir + "/events"
+}
+
+func (r Registry) Create() (*did.Document, error) {
+	doc, err := r.didDocCreator.Create()
+	if err != nil {
+		return nil, fmt.Errorf("could not create did document: %w", err)
+	}
+
+	// Fixme: The doc should not be stored but send to the network.
+	metaData := types.DocumentMetadata{
+		Created: time.Now(),
+		Version: 0,
+	}
+	err = r.store.Write(*doc, metaData)
+	if err != nil {
+		return nil, fmt.Errorf("could not store created did document: %w", err)
+	}
+	return doc, nil
+}
+
+func (r Registry) Resolve(dID did.DID, metadata *types.ResolveMetaData) (*did.Document, *types.DocumentMetadata, error) {
+	return r.store.Resolve(dID, metadata)
+}
+
+func (r Registry) Update(dID did.DID, hash model.Hash, next did.Document, metadata types.DocumentMetadata) error {
+	return r.store.Update(dID, hash, next, metadata)
```

Comment:
I guess we need some more checks in the metadata as well. There might be stuff missing in the RFC for this.

Reply:
We can introduce an ErrInvalidSuccession which can be used when:
Version of a new doc does not start with 0
Version or updatedAt timestamp of a updated doc !> previous

Reply:
Added #12. Lets do this in a next PR when network and VDR are merged.